components/script/dom/permissions.rs

```diff
@@ -0,0 +1,262 @@
+            },
+        };
+        // Step 4.
+        return p;
+    }
+
+    #[allow(unrooted_must_root)]
+    #[allow(unsafe_code)]
+    // https://w3c.github.io/permissions/#dom-permissions-revoke
+    unsafe fn Revoke(&self, cx: *mut JSContext, permissionDesc: *mut JSObject) -> Rc<Promise> {
+        // Step 1.
+        let root_desc = match Permissions::create_descriptor(cx, permissionDesc) {
+            Ok(descriptor) => descriptor,
+            Err(error) => {
+                let p = Promise::new(&self.global());
+                p.reject_error(cx, error);
+                return p;
+            },
+        };
+
+        let status = PermissionStatus::new(&self.global(), &root_desc);
+
+        // Step 2.
+        match root_desc.name {
+            _ => {
+                Permissions::permission_revoke(&root_desc, &status);
+            },
+        };
+
+        // Step 5.
+        return self.Query(cx, permissionDesc);
+    }
+}
+
+impl PermissionAlgorithm for Permissions {
+    type Descriptor = PermissionDescriptor;
+    type Status = PermissionStatus;
+
+    #[allow(unsafe_code)]
+    fn create_descriptor(cx: *mut JSContext,
+                         permission_descriptor_obj: *mut JSObject)
+                         -> Result<PermissionDescriptor, Error> {
+        rooted!(in(cx) let mut property = UndefinedValue());
+        property.handle_mut().set(ObjectValue(permission_descriptor_obj));
+        unsafe {
+            match PermissionDescriptor::new(cx, property.handle()) {
+                Ok(ConversionResult::Success(descriptor)) => Ok(descriptor),
+                Ok(ConversionResult::Failure(error)) => Err(Error::Type(error.into_owned())),
+                Err(_) => Err(Error::Type(String::from(ROOT_DESC_CONVERSION_ERROR))),
+            }
+        }
+    }
+
+    // https://w3c.github.io/permissions/#boolean-permission-query-algorithm
+    fn permission_query(_cx: *mut JSContext,
+                        _promise: &Rc<Promise>,
+                        _descriptor: &PermissionDescriptor,
+                        status: &PermissionStatus) {
+        // Step 1.
+        status.set_state(get_descriptor_permission_state(status.get_query(), None));
+    }
+
+    // https://w3c.github.io/permissions/#boolean-permission-request-algorithm
+    fn permission_request(cx: *mut JSContext,
+                          promise: &Rc<Promise>,
+                          descriptor: &PermissionDescriptor,
+                          status: &PermissionStatus) {
+        // Step 1.
+        Permissions::permission_query(cx, promise, descriptor, status);
+
+        // TODO: Step 2 - 4: `environment settings object` is not implemented in Servo yet.
+        // For this reason in the `get_descriptor_permission_state` function we can't decide
+        // if we have a secure context or not, or store the previous invocation results.
+        // Without these the remaining steps can't be implemented properly.
```

Comment:
We can use `GlobalScope::current()` to give us the equivalent of the `current environment settings object`.